---
title: "MT_foreign_donor_swap_for_github_V2"
author: "Laura Brettell"
date: "2023-01-04"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### About

An experiment was carried out to determine how mosquitoes respond to a microbiome transplant. Transplants were performed using microbiomes from donors belonging to the same, or different mosquito species, collected from different environments (lab vs field). The recipient was always Aedes aegypti GALV laboratory-reared mosquitoes, reared in axenic conditions. A conventionally reared treatment was also included.  

# Input required packages

```{r, eval=TRUE, message=FALSE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#        install.packages("BiocManager")
#BiocManager::install("DESeq2")
#if (!requireNamespace("BiocManager", quietly=TRUE))
#  install.packages("BiocManager")
#BiocManager::install("ComplexHeatmap")
#install.packages("UpSetR")
#install.packages("pheatmap")
```

```{r, eval=TRUE, message=FALSE}
library(DESeq2)
library(ggplot2)
library(UpSetR)
library(dplyr) # for %>%
library(tibble) # for tibble::rownames_to_column(df, "geneID")
library(ComplexHeatmap) # for extracting lists from upsetR and alternative heatmap plotting
library(ggrepel) # when overlaying text on volcano plots this makes then not overlap
library(readr) # for read_tsv
library(cowplot) # for plotting side by side
library(circlize)
```


# Input data and filter

These data are in the form of a raw counts file Shannon generated using HiSat2 and an associated metadata file. DEseq2 normalises to account for differences in library size and to reduce the effect of very highly expressed genes(?), and genes with zero or near-zero counts across all the data, so no prior normalisation is required. 

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, in doing so we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. 

Here we are doing one initial filtering step here, where we will keep only genes/rows which show >=10 reads in every replicate of at least one treatment.

```{r}
data <- read.csv("raw_counts.csv", row.names = 1) #import the count data

#Perform the subsetting we did previously, but limited to only the columns that correspond to the experiment
#This results in seven vectors of logical arguments corresponding to whether the gene (row) is greater than 10 counts in all replicates for each condition (or microbiome swap)
data.AegF.L_GB_p <- apply(data[, 1:3], MARGIN = 1, function(x) all(x >= 10))
data.AegL.L_GB_p <- apply(data[, 4:6], MARGIN = 1, function(x) all(x >= 10))
data.AetF.AegL_GB_p <- apply(data[, 7:9], MARGIN = 1, function(x) all(x >= 10))
data.AetL.AegL_GB_p <- apply(data[, 10:12], MARGIN = 1, function(x) all(x >= 10))
data.Ag.Aeg_GB_p <- apply(data[, 13:15], MARGIN = 1, function(x) all(x >= 10))
data.CONV_p <- apply(data[, 16:18], MARGIN = 1, function(x) all(x >= 10))
data.Cu.Aeg_GB.P <- apply(data[, 19:21], MARGIN = 1, function(x) all(x >= 10))

 
#Merge the different logical vectors together
test.logical <- as.data.frame(rbind(data.AegF.L_GB_p, data.AegL.L_GB_p, data.AetF.AegL_GB_p, data.AetL.AegL_GB_p, data.Ag.Aeg_GB_p, data.CONV_p, data.Cu.Aeg_GB.P))

# we want this transposed - 
# we want a dataframe with 7 columns, one column for each condition
test.logical.2 <- as.data.frame(t(test.logical))

#At this point, at least one column in the row must be TRUE, which will indicate that all replicates in that condition have a count greater than 10
#We then get another vector of logical arguments, checking for the above
test.logical.3 <- rowSums(test.logical.2) >=1

#Finally, filter the original dataset with this final set of logicals
data.filtered <- data[test.logical.3, ]
View(data.filtered)
# reduced from 19804 50 9723 genes
```


```{r}

coldata <- read.csv("coldata_expanded_meta.csv", row.names = 1) #import sample data
coldata <- coldata[,c("Microbiome.swap", "Direction", "microbiome.origin", "donor.env")]
coldata$Microbiome.swap <- factor(coldata$Microbiome.swap)
coldata$Direction <- factor(coldata$Direction)
coldata$microbiome.origin <- factor(coldata$microbiome.origin)
coldata$donor.env <- factor(coldata$donor.env)
head(coldata)
```


# preparing the DESeqDataSet object 

Now we have the required data, we need to put it into an object of the required format, a DESeqDataSet (dds)


```{r}
dds <- DESeqDataSetFromMatrix(countData = data.filtered,
                              colData = coldata,
                              design = ~ Direction)
dds
```


# Analysing the within-swap data


Given the conventional treatment is different biologically, we will now remove it from analyses and use the Aeg Lab microbiome donor/Aeg Lab host treatment as the control. As our last analysis compared everything to conventional, we need to rerun the Differential Expression calculations comparing to the new AegL.L baseline. While it shouldn't make any difference leaving the conventional treatment in here (I don't think), I will remove it before any further analyses just in case it interfered with anything else later.

## subsetting the data

```{r}
# subsetting the data to remove the conv treatment
dds.no.conv <- dds[ , dds$Microbiome.swap == "Yes"]
# selecting for all those saples where a swap was done (only conv doesn't fit with this)
colData(dds.no.conv)
# check the number of samples is correct and conv has gone.
```

## rerunning the differential expression analysis

```{r}
dds.no.conv$Direction <-  factor(dds.no.conv$Direction, levels = c("AegF.L", "AetF.AegL", "AetL.AegL", "Ang.Aeg", "Cu.Aeg", "AegL.L")) 
# resetting the baseline to AegL.L (from Conv as it was before)
# here I've switched AegL.L to be last, but it didn't work when I didn't set levels? Perhaps it was because I set the ref as conv before and now it's gone? I didn't think I did, but will continue for now.
DE.no.conv <- DESeq(dds.no.conv)
```


## extracting the pairwise results

```{r}
res.swap.AegF <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "AegF.L", "AegL.L")))
res.swap.AetF <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "AetF.AegL", "AegL.L")))
res.swap.AetL <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "AetL.AegL", "AegL.L")))
res.swap.Ang <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "Ang.Aeg", "AegL.L")))
res.swap.Cu <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "Cu.Aeg", "AegL.L")))
res.swap.AegF <- res.swap.AegF[abs(res.swap.AegF$log2FoldChange)  >= 1.5, ]
res.swap.AetF <- res.swap.AetF[abs(res.swap.AetF$log2FoldChange)  >= 1.5, ]
res.swap.AetL <- res.swap.AetL[abs(res.swap.AetL$log2FoldChange)  >= 1.5, ]
res.swap.Ang <- res.swap.Ang[abs(res.swap.Ang$log2FoldChange)  >= 1.5, ]
res.swap.Cu <- res.swap.Cu[abs(res.swap.Cu$log2FoldChange)  >= 1.5, ]
genes.res.swap.AegF <- rownames(na.omit(res.swap.AegF[res.swap.AegF$padj < 0.05, ]))
# 447 genes
genes.res.swap.AetF <- rownames(na.omit(res.swap.AetF[res.swap.AetF$padj < 0.05, ]))
# 448 genes
genes.res.swap.AetL <- rownames(na.omit(res.swap.AetL[res.swap.AetL$padj < 0.05, ]))
# 55 genes
genes.res.swap.Ang <- rownames(na.omit(res.swap.Ang[res.swap.Ang$padj < 0.05, ]))
# 19 genes
genes.res.swap.Cu <- rownames(na.omit(res.swap.Cu[res.swap.Cu$padj < 0.05, ]))
# 49 genes
# and stitch the lists
all.comps.to.AegL <- list(genes.res.swap.AegF, genes.res.swap.AetF, genes.res.swap.AetL, genes.res.swap.Ang, genes.res.swap.Cu)
# put the names on in a nice way
names(all.comps.to.AegL) <- c("Ae. aegypti (field)", "Ae. taeniorhynchus (field)", "Ae. taeniorhynchus (lab)", "An. gambiae (lab)", "Cx. tarsalis (lab)")
```

(it is convenient here to extract lists of genes with padj and log2fold change for the pairwise comps of the field caught swaps to control - we will need these later for volcano plots)... I don't think I actually need this


```{r}
# first AegL
#res.swap.AegF_new <- res.swap.AegF  # already has foldchange cutoff
#res.swap.AegF_new <- na.omit(res.swap.AegF_new[res.swap.AegF_new$padj < 0.05, ])
# this is 246 genes
#write.csv(res.swap.AegF_new, file = "./res.swap.AegF_new.csv", row.names = TRUE)
# then AetL
#res.swap.AetF_new <- res.swap.AetF  # already has foldchange cutoff
#res.swap.AetF_new <- na.omit(res.swap.AetF_new[res.swap.AetF_new$padj < 0.05, ])
# this is 260 genes
#write.csv(res.swap.AetF_new, file = "./res.swap.AetF_new.csv", row.names = TRUE)
```


Here I will make an upset plot to look at whether there are genes commonly affected when comparing to AegLab-lab baseline.


```{r}
# alternative colour upset plot
upsetplot2 <- UpSetR::upset(fromList(all.comps.to.AegL), 
                order.by = "freq",
                sets.bar.color =      c("#2b9b81","#2b9b81","#f4c40f","#f4c40f","#f4c40f"))
             
             
# trying to highlight bars, or at least dots containing the field asociated genes but so far having no luck!    - not sure I need to            
        
              
upsetplot2
```



Now this shows an interesting pattern, that there aren't very many statistically significantly enriched genes when comparing microbiomes swaps received from lab donors, regardless of species, but when the donor is a field mosquito there are more differentially expressed genes, and the most are when it's a field mosquito of a different species.


# extracting lists of interesting genes

For this upset plot, we have no genes which are diff expressed in all comparisons to the swap where the recipient received her own microbiome, which makes sense really, especially as we know there is no immune signal.

Instead, here I will extract genelists for the 3 biggest bars and investigate functions of those. These are the 2 field donors, individually and the shared genes.
        
```{r}
# first, create a matrix of counts from the upsetR plot
comb_mat_to_conv_2 <- make_comb_mat(all.comps.to.AegL)
comb_mat_to_conv_2
```


```{r}
# then we want to extract the genes in our intersect of interest by calling them using binary codes
#Pulls out the set containing all items ie all 1's
shared_genes_field_to_AegL <- extract_comb(comb_mat_to_conv_2, "11000") 
# check this is the right number - yes 136
# and now for the other 2 lists
genes_AegF_to_AegL <- extract_comb(comb_mat_to_conv_2, "10000")
# 285 genes
genes_AetF_to_AegL <- extract_comb(comb_mat_to_conv_2, "01000")
# 270 genes
```




# Now to look at those genes which show up only in the comparisons where the donor was field derived (n=136) - ie in both the AetF and AegF comparisons



Doing a heatmap first to look at the directions of the change ie are they the same in both comparisons

Firstly, I will make a dataframe of all IDs passing the thresholds (ie those which went into the upset plot), showing their fold changes compared to AegL.L control, using the same method as I did earlier:



```{r}
# generate vector of all geneIDs to be used
genes.res.swap.all <- genes.res.swap.AegF
genes.res.swap.all <- append(genes.res.swap.all, genes.res.swap.AetF)
genes.res.swap.all <- append(genes.res.swap.all, genes.res.swap.AetL)
genes.res.swap.all <- append(genes.res.swap.all, genes.res.swap.Ang)
genes.res.swap.all <- append(genes.res.swap.all, genes.res.swap.Cu)
#1018 genes
genes.res.swap.dupl.removed <- genes.res.swap.all[!duplicated(genes.res.swap.all)]
summary(genes.res.swap.dupl.removed)
#803 genes after removing duplicates
```


Then collate the fold changes of all these genes compared to the AegL.L control


```{r}
# extract the results again, to keep those genes identified as being significantly up/down regulated in any pairwise comp
res.swap.AegF.all <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "AegF.L", "AegL.L")))
res.swap.AetF.all <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "AetF.AegL", "AegL.L")))
res.swap.AetL.all <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "AetL.AegL", "AegL.L")))
res.swap.Ang.all <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "Ang.Aeg", "AegL.L")))
res.swap.Cu.all <- as.data.frame(results(DE.no.conv, contrast = c("Direction", "Cu.Aeg", "AegL.L")))
foldchange.swap.AegF <- res.swap.AegF.all  %>% select(log2FoldChange)
foldchange.swap.AetF <- res.swap.AetF.all  %>% select(log2FoldChange)
foldchange.swap.AetL <- res.swap.AetL.all  %>% select(log2FoldChange)
foldchange.swap.Ang <- res.swap.Ang.all  %>% select(log2FoldChange)
foldchange.swap.Cu <- res.swap.Cu.all  %>% select(log2FoldChange)
colnames(foldchange.swap.AegF) <- c("AegF")
colnames(foldchange.swap.AetF) <- c("AetF")
colnames(foldchange.swap.AetL) <- c("AetL")
colnames(foldchange.swap.Ang) <- c("Ang")
colnames(foldchange.swap.Cu) <- c("Cu")
foldchange.swap.AegF <- tibble::rownames_to_column(foldchange.swap.AegF, "geneID")
foldchange.swap.AetF <- tibble::rownames_to_column(foldchange.swap.AetF, "geneID")
foldchange.swap.AetL <- tibble::rownames_to_column(foldchange.swap.AetL, "geneID")
foldchange.swap.Ang <- tibble::rownames_to_column(foldchange.swap.Ang, "geneID")
foldchange.swap.Cu <- tibble::rownames_to_column(foldchange.swap.Cu, "geneID")
bound.swap <-merge(foldchange.swap.AegF, foldchange.swap.AetF, by="geneID")
bound.swap <-merge(bound.swap, foldchange.swap.AetL, by="geneID")
bound.swap <-merge(bound.swap, foldchange.swap.Ang, by="geneID")
bound.swap <-merge(bound.swap, foldchange.swap.Cu, by="geneID")
head(bound.swap)
# now to re-set the row names as geneIDs
rownames(bound.swap) <- bound.swap$geneID
# and subset the results to keep only the genes of interest
bound.swap <- bound.swap[genes.res.swap.dupl.removed, ]
# and remove the redundant column 'geneID'
bound.swap <- bound.swap[c(2:6)]
```

Make heatmap of all comps to conv
```{r}
col_fun = colorRamp2(c(-10, -1.5, -1.499, 1.499, 1.5, 10), c( "#283593", "#C5CAE9","snow2", "snow2", "#FFE0B2", "#E65100"))
col_fun(seq(-10, 10))
# boundaries were chosen to span to the max and min and leave those which were on the middle (not statistically signif as grey)

bound.swap[is.na(bound.swap)] <- 0 # NAs need to be zeros with this heatmap package. We have some NAs that were in the correct fold change ranges but not statisticlly signif


heatmap_bound.swap <- Heatmap(bound.swap, name = "mat", col = col_fun, row_names_gp = gpar(fontsize = 3))
heatmap_bound.swap

heatmap_bound.swap <- Heatmap(bound.swap, name = "mat", col = col_fun, show_row_names=F)
heatmap_bound.swap


```
Make a spreadsheet of all fold changes to conv with NAs in the place of all cells where padj is >0.05


```{r}
foldchange.padj.swap.AetF <- res.swap.AetF.all  %>% select(log2FoldChange, padj)
foldchange.padj.swap.AetL <- res.swap.AetL.all  %>% select(log2FoldChange, padj)
foldchange.padj.swap.AegF <- res.swap.AegF.all  %>% select(log2FoldChange, padj)
foldchange.padj.swap.Ang <- res.swap.Ang.all  %>% select(log2FoldChange, padj)
foldchange.padj.swap.Cu <- res.swap.Cu.all  %>% select(log2FoldChange, padj)

colnames(foldchange.padj.swap.AetF) <- c("AetF", "padjAetF")
colnames(foldchange.padj.swap.AetL) <- c("AetL", "padjAetL")
colnames(foldchange.padj.swap.AegF) <- c("AegF", "padjAegF")
colnames(foldchange.padj.swap.Ang) <- c("Ang", "padjAng")
colnames(foldchange.padj.swap.Cu) <- c("Cu", "padjCu")

foldchange.padj.swap.AetF <- tibble::rownames_to_column(foldchange.padj.swap.AetF, "geneID")
foldchange.padj.swap.AetL <- tibble::rownames_to_column(foldchange.padj.swap.AetL, "geneID")
foldchange.padj.swap.AegF <- tibble::rownames_to_column(foldchange.padj.swap.AegF, "geneID")
foldchange.padj.swap.Ang <- tibble::rownames_to_column(foldchange.padj.swap.Ang, "geneID")
foldchange.padj.swap.Cu <- tibble::rownames_to_column(foldchange.padj.swap.Cu, "geneID")

foldchange.padj.swap.AetF <- foldchange.padj.swap.AetF %>% mutate(AetF = ifelse(padjAetF >= 0.05, "NA", AetF))
foldchange.padj.swap.AetL <- foldchange.padj.swap.AetL %>% mutate(AetL = ifelse(padjAetL >= 0.05, "NA", AetL))
foldchange.padj.swap.AegF <- foldchange.padj.swap.AegF %>% mutate(AegF = ifelse(padjAegF >= 0.05, "NA", AegF))
foldchange.padj.swap.Ang <- foldchange.padj.swap.Ang %>% mutate(Ang = ifelse(padjAng >= 0.05, "NA", Ang))
foldchange.padj.swap.Cu <- foldchange.padj.swap.Cu %>% mutate(Cu = ifelse(padjCu >= 0.05, "NA", Cu))


#then combine
bound.padj.swap <-merge(foldchange.padj.swap.AetF, foldchange.padj.swap.AetL, by="geneID")
bound.padj.swap <-merge(bound.padj.swap, foldchange.padj.swap.AegF, by="geneID")
bound.padj.swap <-merge(bound.padj.swap, foldchange.padj.swap.Ang, by="geneID")
bound.padj.swap <-merge(bound.padj.swap, foldchange.padj.swap.Cu, by="geneID")

head(bound.padj.swap)

# now to re-set the row names as geneIDs
rownames(bound.padj.swap) <- bound.padj.swap$geneID

# remove the redundant column 'geneID' and the padj columns
bound.padj.swap <- bound.padj.swap[c(2,4,6,8,10)]

# and subset the results to keep only the genes of interest
bound.padj.swap <- bound.padj.swap[genes.res.swap.dupl.removed, ]

# write a csv of these results for supp

write.csv(bound.padj.swap, "all_comps_to_original_transplant.csv")


```


Looking at just the field swaps

```{r}
# subset to keep only genes of interest (those in field comps)
bound.swap.shared.field <- bound.swap[shared_genes_field_to_AegL, c("AegF", "AetF") ]

heatmap_bound.swap.shared.field <- Heatmap(bound.swap.shared.field, name = "mat", col = col_fun, row_names_gp = gpar(fontsize = 3))
heatmap_bound.swap.shared.field

```


Create heatmap in another way - colours a bit less intuitive

```{r heatmap_136_shared_field_to_AegL.L}
# create heatmap
heatmap_136_shared_field_to_AegL.L <- pheatmap(bound.swap.shared.field,  # matrix of counts
         show_rownames = TRUE,
         fontsize_row = 2,
         border_color = NA,
         legend = TRUE,
         cluster_cols = FALSE, 
         cluster_rows = TRUE,
         scale = 'none',
         color = colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")))(100)
        )
heatmap_136_shared_field_to_AegL.L
```


This shows that the direction of change is nearly always the same in these shared genes.


*******************************************************************************
# heatmap of all genes in pairwise comp between field microbiome transplants and original microbiome transplant control

```{r}
# I previously extracted the results from each pairwise comparison to AegL.L
# Here I will take the log2foldchange column AND the padj column
foldchange.padj.swap.AegF <- res.swap.AegF.all  %>% select(log2FoldChange, padj)
# rename the column to the sample name 
colnames(foldchange.padj.swap.AegF) <- c("AegF", "padjAegF")
# give the geneIDs their own column, to be able to merge later
foldchange.padj.swap.AegF <- tibble::rownames_to_column(foldchange.padj.swap.AegF, "geneID")
# if padj >0.05, replace the value in the fold change column (now AegF) with NA (later rows will remain for non-signif values - if the value was signif in another comparison, so these need replacing with NAs). Doing this in 3 parts is not efficient but works.
foldchange.padj.swap.AegF <- foldchange.padj.swap.AegF %>% mutate(AegF = ifelse(padjAegF >= 0.05, "NA", AegF))


# and AetF
foldchange.padj.swap.AetF <- res.swap.AetF.all  %>% select(log2FoldChange, padj)
colnames(foldchange.padj.swap.AetF) <- c("AetF", "padjAetF")
foldchange.padj.swap.AetF <- tibble::rownames_to_column(foldchange.padj.swap.AetF, "geneID")
foldchange.padj.swap.AetF <- foldchange.padj.swap.AetF %>% mutate(AetF = ifelse(padjAetF >= 0.05, "NA", AetF))


bound.field.padj <- merge(foldchange.padj.swap.AegF, foldchange.padj.swap.AetF, by="geneID")

rownames(bound.field.padj) <- bound.field.padj$geneID
bound.field.padj <- bound.field.padj[c(2,4)]

bound.field.padj <- as.data.frame(apply(bound.field.padj, 2, as.numeric), row.names = rownames(bound.field.padj)) 

all.field.swap.genes <- genes.res.swap.AegF
all.field.swap.genes <- append(all.field.swap.genes, genes.res.swap.AetF)
all.field.swap.genes <- all.field.swap.genes[!duplicated(all.field.swap.genes)]

bound.field.padj <- bound.field.padj[all.field.swap.genes,]
```


```{r}
col_fun = colorRamp2(c(-10, -1.5, -1.499, 1.499, 1.5, 4), c( "#283593", "#C5CAE9","snow2", "snow2", "#FFE0B2", "#E65100"))
col_fun(seq(-10, 4))

bound.field.padj[is.na(bound.field.padj)] <- 0 # NAs need to be zeros with this heatmap package


heatmap_bound.field.padj_rownames <- Heatmap(bound.field.padj, name = "mat", col = col_fun, row_names_gp = gpar(fontsize = 3))
heatmap_bound.field.padj_rownames

heatmap_bound.field.padj <- Heatmap(bound.field.padj, name = "mat", col = col_fun, show_row_names=F)
heatmap_bound.field.padj




heatmap_bound.field.padj <- Heatmap(bound.field.padj, col = col_fun)
heatmap_bound.field.padj
```





********************************************************************************

# Extract data from DE genes when comparing field-derived transplants to original microbiome transplant controls.



```{r}
res.swap.AegF.all.copy <- res.swap.AegF.all %>% select(log2FoldChange, padj)
res.swap.AegF.all.copy <- res.swap.AegF.all.copy[genes.res.swap.AegF,]
write.csv(res.swap.AegF.all.copy, "res.swap.AegF.all.copy.csv")
res.swap.AetF.all.copy <- res.swap.AetF.all %>% select(log2FoldChange, padj)
res.swap.AetF.all.copy <- res.swap.AetF.all.copy[genes.res.swap.AetF,]
write.csv(res.swap.AetF.all.copy, "res.swap.AetF.all.copy.csv")

```

Then use these resulting lists (separated by up and down) as inputs for Vectorbase GO enrichment



Write other comps too for supp data
```{r}
res.swap.Ang.all.copy <- res.swap.Ang.all %>% select(log2FoldChange, padj)
res.swap.Ang.all.copy <- res.swap.Ang.all.copy[genes.res.swap.Ang,]
write.csv(res.swap.Ang.all.copy, "res.swap.Ang.all.copy.csv")

res.swap.AetL.all.copy <- res.swap.AetL.all %>% select(log2FoldChange, padj)
res.swap.AetL.all.copy <- res.swap.AetL.all.copy[genes.res.swap.AetL,]
write.csv(res.swap.AetL.all.copy, "res.swap.AetL.all.copy.csv")

res.swap.Cu.all.copy <- res.swap.Cu.all %>% select(log2FoldChange, padj)
res.swap.Cu.all.copy <- res.swap.Cu.all.copy[genes.res.swap.Cu,]
write.csv(res.swap.Cu.all.copy, "res.swap.Cu.all.copy.csv")

```



