---
title: "MT_RNASeq_for_Github"
author: "Laura Brettell"
date: "08/08/2022"
output: html_document
keep_md: yes
---

#### About

An experiment was carried out to determine how mosquitoes respond to a microbiome transplant. Transplants were performed using microbiomes from donors belonging to the same, or different mosquito species, collected from different environments (lab vs field). The recipient was always Aedes aegypti GALV laboratory-reared mosquitoes, reared in axenic conditions. A conventionally reared treatment was also included.  



# Input required packages

```{r, eval=TRUE, message=FALSE}

#if (!requireNamespace("BiocManager", quietly = TRUE))
#        install.packages("BiocManager")
#BiocManager::install("DESeq2")


#if (!requireNamespace("BiocManager", quietly=TRUE))
#  install.packages("BiocManager")
#BiocManager::install("ComplexHeatmap")

#install.packages("UpSetR")
#install.packages("pheatmap")

```

```{r, eval=TRUE, message=FALSE}

library(DESeq2)
library(ggplot2)
library(UpSetR)
library(dplyr) # for %>%
library(tibble) # for tibble::rownames_to_column(df, "geneID")
library(ComplexHeatmap) # for extracting lists from upsetR and alternative heatmap plotting
library(ggrepel) # when overlaying text on volcano plots this makes then not overlap
library(readr) # for read_tsv
library(cowplot) # for plotting side by side
library(circlize) # for setting up colours in complexheatmaps

```


# Input data

These data are in the form of a raw counts file Shannon generated using HiSat2 and an associated metadata file. DEseq2 normalises to account for differences in library size and to reduce the effect of very highly expressed genes(?), and genes with zero or near-zero counts across all the data, so no prior normalisation is required. 

```{r}

raw_counts <- read.csv("raw_counts.csv", row.names = 1) #import the count data
head(raw_counts)

coldata <- read.csv("coldata_expanded_meta.csv", row.names = 1) #import sample data
coldata <- coldata[,c("Microbiome.swap", "Direction", "microbiome.origin", "donor.env")]

coldata$Microbiome.swap <- factor(coldata$Microbiome.swap)
coldata$Direction <- factor(coldata$Direction)
coldata$microbiome.origin <- factor(coldata$microbiome.origin)
coldata$donor.env <- factor(coldata$donor.env)


head(coldata)

```

# Prefiltering

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, in doing so we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. 

Here I am filtering to rows which have at least 10 counts in every sample.

```{r}

filtered_counts <- raw_counts[apply(raw_counts[, -1], MARGIN = 1, function(x) all(x >= 10)), ]
# this reduces the number of rows from 19804 to 7836

```

# preparing the DESeq2 object 

Now we have the required data, we need to put it into an object of the required format, a DESeqDataSet (dds)


```{r}
dds <- DESeqDataSetFromMatrix(countData = filtered_counts,
                              colData = coldata,
                              design = ~ Direction)

dds
```

Now to prepare for the Differential expression analysis, we ensure the reference level is set appropriately ie is the baseline to which all others will be compared (in our case the conventional treatment - as our initial analysis will look at the effect of just doing a swap). This is all still essentially metadata at this point, no comparisons have been made yet. 


```{r}
dds$Direction <- factor(dds$Direction, levels = c("AegL.L", "AegF.L", "AetF.AegL", "AetL.AegL", "Ang.Aeg", "Cu.Aeg", "Conventional"))

```



# Looking at the Differential Expression of each of the swaps individually compared to the Conventional treatment

Just doing a swap is quite a big change so it will be good to look at all those genes affected by this.

Firstly, I need to perform the differential expression analysis, with the variable being tested set to 'Direction' ie all the different sample types (this was set up earlier when we set up the dds object). 

I will then go on to creating a heatmap of fold changes across the set and an upset plot to identify genes which show up in multiple comparisons.


# running the differential expression analysis


```{r}
DE.dds <- DESeq(dds)

```
Now have a quick look at the results

```{r}

res.all <- results(DE.dds)

res.all
# by default this just shows the last comparison that R performed (alphabetically. 

```


# Extracting the results of interest and filtering

Firstly, extract all the results of interest ie every pairwise comparison to Conventional and apply cutoffs to remove those results which either didn't show much of a change, or which had a low probability of being correctly called. For this I will retain only the geneIDs in each pairwise Differential expression result with an adjusted p value of < 0.05 and a log2 fold change of 1.5.




```{r}

# I tried to apply the cutoffs to the results as a whole but got errors about the presence of NAs and was unable to fully remove those with na.omit (for some), so I will extract each pairwise result first, then apply the filters individually. 

# so extract the results of the first pairwise comparison as a dataframe
res.all.AegL.L.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AegL.L", "Conventional")))
# 7836 geneIDs

# filter by all those geneIDs with a log2 fold change of > 1.5 or <- 1.5
res.all.AegL.L.vs.conv <- res.all.AegL.L.vs.conv[abs(res.all.AegL.L.vs.conv$log2FoldChange) >= 1.5, ]
# 433 geneIDs

# then filter by padj < 0.05, remove NAs (introduced during the initial filtering), and pull out the geneIDs
gene.IDs.AegL.L.vs.conv <- rownames(na.omit(res.all.AegL.L.vs.conv[res.all.AegL.L.vs.conv$padj < 0.05, ]))
# now is 1372 geneIDs with na's omitted 


summary(gene.IDs.AegL.L.vs.conv)
```
Then repeat this for the rest of the pairwise comparisons (not the most elegant way of doing this!)

```{r}
res.all.AegF.L.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AegF.L", "Conventional")))
res.all.AegF.L.vs.conv <- res.all.AegF.L.vs.conv[abs(res.all.AegF.L.vs.conv$log2FoldChange)  >= 1.5, ]
gene.IDs.AegF.L.vs.conv <- rownames(na.omit(res.all.AegF.L.vs.conv[res.all.AegF.L.vs.conv$padj < 0.05, ]))
# 312 genes


res.all.AetF.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AetF.AegL", "Conventional")))
res.all.AetF.vs.conv <- res.all.AetF.vs.conv[abs(res.all.AetF.vs.conv$log2FoldChange)  >= 1.5, ]
gene.IDs.AetF.L.vs.conv <- rownames(na.omit(res.all.AetF.vs.conv[res.all.AetF.vs.conv$padj < 0.05, ]))  
#85  genes


res.all.AetL.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AetL.AegL", "Conventional")))
res.all.AetL.vs.conv <- res.all.AetL.vs.conv[abs(res.all.AetL.vs.conv$log2FoldChange)  >= 1.5, ]
gene.IDs.AetL.L.vs.conv <- rownames(na.omit(res.all.AetL.vs.conv[res.all.AetL.vs.conv$padj < 0.05, ])) 
# 420 genes


res.all.Ang.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "Ang.Aeg", "Conventional")))
res.all.Ang.vs.conv <- res.all.Ang.vs.conv[abs(res.all.Ang.vs.conv$log2FoldChange)  >= 1.5, ]
gene.IDs.Ang.vs.conv <- rownames(na.omit(res.all.Ang.vs.conv[res.all.Ang.vs.conv$padj < 0.05, ]))
# 307 genes


res.all.Cu.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "Cu.Aeg", "Conventional")))
res.all.Cu.vs.conv <- res.all.Cu.vs.conv[abs(res.all.Cu.vs.conv$log2FoldChange)  >= 1.5, ]
gene.IDs.Cu.vs.conv <- rownames(na.omit(res.all.Cu.vs.conv[res.all.Cu.vs.conv$padj < 0.05, ]))
# 286 genes



# and stitch the lists
all.comps.to.conv <- list(gene.IDs.AegL.L.vs.conv, gene.IDs.AegF.L.vs.conv, gene.IDs.AetF.L.vs.conv, gene.IDs.AetL.L.vs.conv, gene.IDs.Ang.vs.conv, gene.IDs.Cu.vs.conv)


names(all.comps.to.conv) <- c("gene.IDs.AegL.L.vs.conv", "gene.IDs.AegF.L.vs.conv", "gene.IDs.AetF.L.vs.conv", "gene.IDs.AetL.L.vs.conv", "gene.IDs.Ang.vs.conv", "gene.IDs.Cu.vs.conv")


```


Then compiling all geneIDs of interest ie all those genes that pass the threshold in at least one pairwise comparison. These lists are then concatenated and duplicates removed to obtain a full geneID list of interest. 

These genes will be those that are looked at in the first heatmap and upset plot.

```{r}
# first to create a vector containing all the genes to be used 

all.gene.IDs <- gene.IDs.AegL.L.vs.conv
all.gene.IDs <- append(all.gene.IDs, gene.IDs.AegF.L.vs.conv)
all.gene.IDs <- append(all.gene.IDs, gene.IDs.AetF.L.vs.conv)
all.gene.IDs <- append(all.gene.IDs, gene.IDs.AetL.L.vs.conv)
all.gene.IDs <- append(all.gene.IDs, gene.IDs.Ang.vs.conv)
all.gene.IDs <- append(all.gene.IDs, gene.IDs.Cu.vs.conv)

# this vector contains 1782 genes,some of which are duplicates (ie feature in multiple lists), so need removing. 


all.gene.IDs.dupl.removed <- all.gene.IDs[!duplicated(all.gene.IDs)]
summary(all.gene.IDs.dupl.removed)
# now there are 876 unique genes for inclusion in the heatmap - the same as before

```


# Upset plot to visualise which DE genes are common to multpile comparisons

Now to just consider which genes were identified in each comparison and whether the same genes were identified as statistically significant up/down regulated (according to the parameters I set). (forgetting about the direction/scale of the change for now).


```{r first_upset_plot}
# duplicating so i can try to make the names of the sets nicer for the plots, but still keeping the original just in case
all.comps.to.conv.renamed <- all.comps.to.conv
names(all.comps.to.conv.renamed) # to check the order

# make names nicer
names(all.comps.to.conv.renamed) <- c("Ae. aegypti (lab)", "Ae. aegypti (field)", "Ae. taeniorhynchus (field)", "Ae. taeniorhynchus (lab)", "An. gambiae (lab)", "Cx. tarsalis (lab)")

upsetplot1 <- UpSetR::upset(fromList(all.comps.to.conv.renamed), 
                order.by = "freq", 
                nsets = 38,
                queries = list(list(query = intersects, params = list(
                  "Ae. aegypti (lab)", 
                  "Ae. aegypti (field)", 
                  "Ae. taeniorhynchus (field)", 
                  "Ae. taeniorhynchus (lab)", 
                  "An. gambiae (lab)", 
                  "Cx. tarsalis (lab)"), 
                  color = "red", active = T)))

upsetplot1


```

There are lots of genes which are only on one, or a subset of comparisons, but there are 18 (in red) which are common to all swaps. 




# investigating the 'swap effect' in more depth

Now to look at the 18 genes that were identified as being enriched in every swap compared to the baseline conventional treatment (above, in red). We don't know if they are enriched in the same direction at all, or whether the swaps show both up and down regulation of these genes compared to conventional


So, firstly to extract the list of 18 geneIDs to assign annotations         
        
```{r}

# first, create a matrix of counts from the upsetR plot

comb_mat_to_conv <- make_comb_mat(all.comps.to.conv) # this is from complexheatmap package
comb_mat_to_conv



```


```{r}

# then we want to extract the genes in our intersect of interest by calling them using binary codes

#Pulls out the set containing all items ie all 1's
shared_genes_to_conv <- extract_comb(comb_mat_to_conv, "111111")


```




# Are these genes showing the same direction of regulation in the different comaprisons? 

This requires extracting all the fold change results from each of the pairwise comparisons for the 18 geneIDs of interest, then collating and creating a heatmap


# extracting and collating results

```{r}

#### add columns to contain the log2 fold change data from each of the pairwise comps used to conv

# this needs to go back to the res.all and extract the comps again because in organising the data for the upsetR genelists, each comp was filtered separately, so some genes will have been removed that I want left in because they are statistically significant in a different pairwise comp


# extract the result
res.full.AegF.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AegF.L", "Conventional")))


# just take the log2foldchange column
foldchange.AegF <- res.full.AegF.vs.conv  %>% select(log2FoldChange)


# rename the column to the sample name (otherwise when I bind, all samples will be indistinguishable)
colnames(foldchange.AegF) <- c("AegF")


# give the geneIDs their own column, to be able to merge later
foldchange.AegF <- tibble::rownames_to_column(foldchange.AegF, "geneID")



# now for the other pairwise comps


res.full.AegL.L.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AegL.L", "Conventional")))
res.full.AetF.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AetF.AegL", "Conventional")))
res.full.AetL.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "AetL.AegL", "Conventional")))
res.full.Ang.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "Ang.Aeg", "Conventional")))
res.full.Cu.vs.conv <- as.data.frame(results(DE.dds, contrast = c("Direction", "Cu.Aeg", "Conventional")))


foldchange.AegL <- res.full.AegL.L.vs.conv  %>% select(log2FoldChange)
foldchange.AetF <- res.full.AetF.vs.conv  %>% select(log2FoldChange)
foldchange.AetL <- res.full.AetL.vs.conv  %>% select(log2FoldChange)
foldchange.Ang <- res.full.Ang.vs.conv  %>% select(log2FoldChange)
foldchange.Cu <- res.full.Cu.vs.conv  %>% select(log2FoldChange)


colnames(foldchange.AegL) <- c("AegL")
colnames(foldchange.AetF) <- c("AetF")
colnames(foldchange.AetL) <- c("AetL")
colnames(foldchange.Ang) <- c("Ang")
colnames(foldchange.Cu) <- c("Cu")

foldchange.AegL <- tibble::rownames_to_column(foldchange.AegL, "geneID")
foldchange.AetF <- tibble::rownames_to_column(foldchange.AetF, "geneID")
foldchange.AetL <- tibble::rownames_to_column(foldchange.AetL, "geneID")
foldchange.Ang <- tibble::rownames_to_column(foldchange.Ang, "geneID")
foldchange.Cu <- tibble::rownames_to_column(foldchange.Cu, "geneID")

bound <-merge(foldchange.AegF, foldchange.AegL, by="geneID")
bound <-merge(bound, foldchange.AetF, by="geneID")
bound <-merge(bound, foldchange.AetL, by="geneID")
bound <-merge(bound, foldchange.Ang, by="geneID")
bound <-merge(bound, foldchange.Cu, by="geneID")
# I don't seem to be able to bind altogether?


head(bound)

# now to re-set the row names as geneIDs
rownames(bound) <- bound$geneID

# remove the redundant column 'geneID'
bound <- bound[c(2:7)]

# and subset the results to keep only the genes of interest
bound.shared.to.conv <- bound[shared_genes_to_conv, ]


```




# make a heatmap of up/down regulation (fold change) from conventional


Now visualising the fold changes from conventional in each sample for each of these 18 genes of interest. 

```{r heatmap_18_genes_to_conv_scale_none}


# create heatmap
heatmap_18_genes_to_conv_scale_none <- pheatmap(
    bound.shared.to.conv,  
    show_rownames = TRUE,
    fontsize_row = 7,
    border_color = NA,
    legend = TRUE,
    cluster_cols = TRUE, 
    cluster_rows = TRUE,
    scale = 'none', 
    color = colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = "RdYlBu")))(100)
  )

heatmap_18_genes_to_conv_scale_none

```


All 18 geneIDs identified as deferentially expressed in all comparisons to conventional, pretty much have consistent directions of change.





# Pull together all transcripts in Upsetplot1 and identify their direction of change
The heatmap of the 18 genes identified in all comparisons showed consistency in the direction of change. Does this pattern hold more generally?


```{r}

# I previously extracted the results from each pairwise comparison to conv
# Here I will take the log2foldchange column AND the padj column
foldchange.padj.AegF <- res.full.AegF.vs.conv  %>% select(log2FoldChange, padj)

# rename the column to the sample name 
colnames(foldchange.padj.AegF) <- c("AegF", "padjAegF")

# give the geneIDs their own column, to be able to merge later
foldchange.padj.AegF <- tibble::rownames_to_column(foldchange.padj.AegF, "geneID")

# if padj >0.05, replace the value in the fold change column (now AegF) with NA (later rows will remain for non-signif values - if the value was signif in another comparison, so these need replacing with NAs). Doing this in 3 parts is not efficient but works.
foldchange.padj.AegF <- foldchange.padj.AegF %>% mutate(AegF = ifelse(padjAegF >= 0.05, "NA", AegF))


```

Struggling to fill NAs for the fold change thresholds, for now will just keep all of them in the dataset and can remove/colour later...


now for the other pairwise comps...

```{r}
foldchange.padj.AetF <- res.full.AetF.vs.conv  %>% select(log2FoldChange, padj)
foldchange.padj.AetL <- res.full.AetL.vs.conv  %>% select(log2FoldChange, padj)
foldchange.padj.AegL <- res.full.AegL.L.vs.conv  %>% select(log2FoldChange, padj)
foldchange.padj.Ang <- res.full.Ang.vs.conv  %>% select(log2FoldChange, padj)
foldchange.padj.Cu <- res.full.Cu.vs.conv  %>% select(log2FoldChange, padj)

colnames(foldchange.padj.AetF) <- c("AetF", "padjAetF")
colnames(foldchange.padj.AetL) <- c("AetL", "padjAetL")
colnames(foldchange.padj.AegL) <- c("AegL", "padjAegL")
colnames(foldchange.padj.Ang) <- c("Ang", "padjAng")
colnames(foldchange.padj.Cu) <- c("Cu", "padjCu")

foldchange.padj.AetF <- tibble::rownames_to_column(foldchange.padj.AetF, "geneID")
foldchange.padj.AetL <- tibble::rownames_to_column(foldchange.padj.AetL, "geneID")
foldchange.padj.AegL <- tibble::rownames_to_column(foldchange.padj.AegL, "geneID")
foldchange.padj.Ang <- tibble::rownames_to_column(foldchange.padj.Ang, "geneID")
foldchange.padj.Cu <- tibble::rownames_to_column(foldchange.padj.Cu, "geneID")

foldchange.padj.AetF <- foldchange.padj.AetF %>% mutate(AetF = ifelse(padjAetF >= 0.05, "NA", AetF))
foldchange.padj.AetL <- foldchange.padj.AetL %>% mutate(AetL = ifelse(padjAetL >= 0.05, "NA", AetL))
foldchange.padj.AegL <- foldchange.padj.AegL %>% mutate(AegL = ifelse(padjAegL >= 0.05, "NA", AegL))
foldchange.padj.Ang <- foldchange.padj.Ang %>% mutate(Ang = ifelse(padjAng >= 0.05, "NA", Ang))
foldchange.padj.Cu <- foldchange.padj.Cu %>% mutate(Cu = ifelse(padjCu >= 0.05, "NA", Cu))


#then combine
bound.padj <-merge(foldchange.padj.AegF, foldchange.padj.AegL, by="geneID")
bound.padj <-merge(bound.padj, foldchange.padj.AetF, by="geneID")
bound.padj <-merge(bound.padj, foldchange.padj.AetL, by="geneID")
bound.padj <-merge(bound.padj, foldchange.padj.Ang, by="geneID")
bound.padj <-merge(bound.padj, foldchange.padj.Cu, by="geneID")

head(bound.padj)

# now to re-set the row names as geneIDs
rownames(bound.padj) <- bound.padj$geneID

# remove the redundant column 'geneID' and the padj columns
bound.padj <- bound.padj[c(2,4,6,8,10,12)]

# and subset the results to keep only the genes of interest
bound.padj <- bound.padj[all.gene.IDs.dupl.removed, ]


```




```{r}


bound.padj.no.nas <- bound.padj

bound.padj.no.nas <- data.frame(apply(bound.padj.no.nas, 2, as.numeric), check.names=, row.names = rownames(bound.padj.no.nas))  # Convert all variable types to numeric, keeping row names as they are
sapply(temp, class)

bound.padj.no.nas[is.na(bound.padj.no.nas)] <- 0 # change all NAs to 0



```

```{r}


col_fun = colorRamp2(c(-10, -1.5, -1.499, 1.499, 1.5, 4), c( "#283593", "#C5CAE9","snow2", "snow2", "#FFE0B2", "#E65100"))
col_fun(seq(-10, 4))


heatmap_876_genes_to_conv_rownames <- Heatmap(bound.padj.no.nas, name = "mat", col = col_fun, row_names_gp = gpar(fontsize = 3))

heatmap_876_genes_to_conv <- Heatmap(bound.padj.no.nas, name = "mat", col = col_fun, show_row_names=F)

heatmap_876_genes_to_conv
```

from the pdf of this plot, then take the row names for all those genes upregulated, downregulated, or a mix
'fold_changes_compared_to_conventional.xlsx'
Then use those genelists as inputs to Vectorbase GO enrichment analysis







